pennylane/optimize/rotosolve.py

```diff
@@ -12,172 +12,411 @@
         """
-        # will single out one variable to change at a time
+        if num_frequencies is None:
+            num_frequencies = [1] * len(args)
+        elif np.isscalar(num_frequencies) and np.isclose(int(num_frequencies), num_frequencies):
+            num_frequencies = [num_frequencies] * len(args)
+        else:
+            if len(num_frequencies) != len(args):
+                raise ValueError(
+                    "The number of the provided numbers of frequencies "
+                    f"({len(num_frequencies)}) does not match the number of function arguments "
+                    f"({len(args)})."
+                )
+
+        if optimizer_kwargs is None:
+            optimizer_kwargs = {}
+
+        if optimizer in [None, "brute"]:
+            optimizer = _brute_optimizer
+            optimizer_kwargs.setdefault("num_steps", 4)
+            optimizer_kwargs.setdefault("Ns", 100)
+        elif optimizer == "shgo":
+            optimizer = _shgo_optimizer
+            optimizer_kwargs.setdefault("bounds", ((-np.pi, np.pi),))
+
+        # will single out one arg to change at a time
         # these hold the arguments not getting updated
         before_args = []
         after_args = list(args)
-
         # mutable version of args to get updated
         args_new = list(args)
 
-        for index, arg in enumerate(args):
-            # removing current arg from after_args
+        # Prepare intermediate minimization results cache
+        if full_output:
+            y_output = []
+        # Compute the very first evaluation in order to be able to cache it
+        H_0 = objective_fn(*args, **kwargs)
+
+        for arg_index, arg in enumerate(list(args)):
+            num_frequency = num_frequencies[arg_index]
             del after_args[0]
 
             if getattr(arg, "requires_grad", True):
                 x_flat = np.fromiter(_flatten(arg), dtype=float)
-
-                # version of objective function that depends on a flattened version of
-                # just the one argument.  All others held constant.
-                objective_fn_flat = lambda x_flat, arg_kw=arg: objective_fn(
-                    *before_args, unflatten(x_flat, arg_kw), *after_args, **kwargs
-                )
-
-                # updating each parameter in current arg
-                for d, _ in enumerate(x_flat):
-                    x_flat = self._rotosolve(objective_fn_flat, x_flat, d)
-
-                args_new[index] = unflatten(x_flat, arg)
+                num_params = len(x_flat)
+                if np.isscalar(num_frequency):
+                    num_frequency_flat = [num_frequency] * num_params
+                else:
+                    num_frequency_flat = np.fromiter(_flatten(num_frequency), dtype=int)
+                    if len(num_frequency_flat) != num_params:
+                        raise ValueError(
+                            "The number of the numbers of frequencies "
```

Comment:
I keep reading this, but 'the number of the numbers' isn't making sense in my head 😆